### ISSUE TITLE
Surface Book 2 : button issues

### ISSUE THREAD

**Issue by deb75** [jakeday/linux-surface#220]
Hello,

I am very happy with the SB2 13" buttons working now.

However, I encounter a few issues : 
- for the volume buttons, I created the files `/etc/acpi/events/surfacebook2-volume-up` and /etc/acpi/events/surfacebook2-volume-down`, as soon as I press the volume buttons, something happens in the Xorg logs :

```
[    37.586] (II) config/udev: Adding input device ACPI Virtual Keyboard Device (/dev/input/event257)
[    37.586] (**) ACPI Virtual Keyboard Device: Applying InputClass "evdev keyboard catchall"
[    37.586] (**) ACPI Virtual Keyboard Device: Applying InputClass "libinput keyboard catchall"
[    37.586] (**) ACPI Virtual Keyboard Device: Applying InputClass "system-keyboard"
[    37.586] (II) Using input driver 'libinput' for 'ACPI Virtual Keyboard Device'
[    37.586] (**) ACPI Virtual Keyboard Device: always reports core events
[    37.586] (**) Option "Device" "/dev/input/event257"
[    37.586] (**) Option "_source" "server/udev"
[    37.587] (II) event257 - ACPI Virtual Keyboard Device: is tagged by udev as: Keyboard
[    37.587] (II) event257 - ACPI Virtual Keyboard Device: device is a keyboard
[    37.587] (II) event257 - ACPI Virtual Keyboard Device: device removed
[    37.593] (**) Option "config_info" "udev:/sys/devices/virtual/input/input54/event257"
[    37.593] (II) XINPUT: Adding extended input device "ACPI Virtual Keyboard Device" (type: KEYBOARD, id 18)
[    37.593] (**) Option "xkb_model" "pc105"
[    37.593] (**) Option "xkb_layout" "fr"
[    37.593] (**) Option "xkb_variant" "oss"
[    37.593] (WW) Option "xkb_options" requires a string value
[    37.593] (II) event257 - ACPI Virtual Keyboard Device: is tagged by udev as: Keyboard
[    37.593] (II) event257 - ACPI Virtual Keyboard Device: device is a keyboard
[   126.437] (II) config/udev: removing device ACPI Virtual Keyboard Device
[   126.437] (II) event257 - ACPI Virtual Keyboard Device: device removed
[   126.443] (II) UnloadModule: "libinput"
[   126.480] (II) config/udev: Adding input device ACPI Virtual Keyboard Device (/dev/input/event257)
[   126.480] (**) ACPI Virtual Keyboard Device: Applying InputClass "evdev keyboard catchall"
[   126.480] (**) ACPI Virtual Keyboard Device: Applying InputClass "libinput keyboard catchall"
[   126.480] (**) ACPI Virtual Keyboard Device: Applying InputClass "system-keyboard"
[   126.480] (II) Using input driver 'libinput' for 'ACPI Virtual Keyboard Device'
[   126.480] (**) ACPI Virtual Keyboard Device: always reports core events
[   126.480] (**) Option "Device" "/dev/input/event257"
[   126.481] (**) Option "_source" "server/udev"
[   126.481] (II) event257 - ACPI Virtual Keyboard Device: is tagged by udev as: Keyboard
[   126.481] (II) event257 - ACPI Virtual Keyboard Device: device is a keyboard
[   126.481] (II) event257 - ACPI Virtual Keyboard Device: device removed
[   126.491] (**) Option "config_info" "udev:/sys/devices/virtual/input/input55/event257"
[   126.491] (II) XINPUT: Adding extended input device "ACPI Virtual Keyboard Device" (type: KEYBOARD, id 18)
[   126.491] (**) Option "xkb_model" "pc105"
[   126.491] (**) Option "xkb_layout" "fr"
[   126.491] (**) Option "xkb_variant" "oss"
[   126.491] (WW) Option "xkb_options" requires a string value
[   126.492] (II) event257 - ACPI Virtual Keyboard Device: is tagged by udev as: Keyboard
[   126.492] (II) event257 - ACPI Virtual Keyboard Device: device is a keyboard
[   164.088] (II) config/udev: removing device ACPI Virtual Keyboard Device
[   164.089] (II) event257 - ACPI Virtual Keyboard Device: device removed
[   164.100] (II) UnloadModule: "libinput"
[   164.120] (II) config/udev: Adding input device ACPI Virtual Keyboard Device (/dev/input/event257)
[   164.120] (**) ACPI Virtual Keyboard Device: Applying InputClass "evdev keyboard catchall"
[   164.120] (**) ACPI Virtual Keyboard Device: Applying InputClass "libinput keyboard catchall"
[   164.120] (**) ACPI Virtual Keyboard Device: Applying InputClass "system-keyboard"
[   164.120] (II) Using input driver 'libinput' for 'ACPI Virtual Keyboard Device'
[   164.120] (**) ACPI Virtual Keyboard Device: always reports core events
[   164.120] (**) Option "Device" "/dev/input/event257"
[   164.120] (**) Option "_source" "server/udev"
[   164.121] (II) event257 - ACPI Virtual Keyboard Device: is tagged by udev as: Keyboard
[   164.121] (II) event257 - ACPI Virtual Keyboard Device: device is a keyboard
[   164.121] (II) event257 - ACPI Virtual Keyboard Device: device removed
[   164.131] (**) Option "config_info" "udev:/sys/devices/virtual/input/input56/event257"
[   164.131] (II) XINPUT: Adding extended input device "ACPI Virtual Keyboard Device" (type: KEYBOARD, id 18)
[   164.131] (**) Option "xkb_model" "pc105"
[   164.131] (**) Option "xkb_layout" "fr"
[   164.131] (**) Option "xkb_variant" "oss"
[   164.131] (WW) Option "xkb_options" requires a string value
[   164.131] (II) event257 - ACPI Virtual Keyboard Device: is tagged by udev as: Keyboard
[   164.131] (II) event257 - ACPI Virtual Keyboard Device: device is a keyboard
```

and then the keyboard and touchpad becomes working strangely : as an example I can no more make key repetitions pressing it. If I launche `acpi_listen`, I am flooded with volume buttons events as if I was continuously pressing them.

But if I reboot without the events files and if I add them after having launched the X session, then everything works fine, including the volume buttons.

- for the power button, pressing it makes the SB2 reboot and not shutdown, I do not manage to understand how the power button works, is it an acpi event as well ? I tried to remove the event, but it still reboots.

 I forgot to mention my OS is debian/buster, original kernel 4.17.3 with latests surface patches.

**Comment by qzed** [comment]
Hmm, I can't reproduce this with a simple event-file writing to a log-file when the buttons are pressed.

Both power and volume buttons are mapped to the respective ACPI events. The only difference is that two keyboard-devices are created (via `gpio_keys`), one for the power button and one for the volume buttons. The two separate devices are needed because the volume buttons should automatically repeat when pressed for a longer time, whereas the power button should not.

**Comment by deb75** [comment]
Hi,

Thanks for your reply.

I removed the event files and volume buttons are working without them. I thought at first one needed to create the appropriate event files pointing to this script (for volume up) : 
```
#!/bin/sh

test -f /usr/share/acpi-support/key-constants || exit 0

. /usr/share/acpi-support/key-constants
acpi_fakekey $KEY_VOLUMEUP
```

So it is not ok for volume buttons, but for the power button, it still reboots after shuttting down, or more precisely, it shuts down completely, but automatically powers on after a few seconds.

**Comment by qzed** [comment]
The buttons should work without any modifications (both power and volume), there are no extra scripts needed.

I just tested this script and basically it creates an ACPI event with source `button/volumeup`, the same as the button driver. Thus with the script you create a recursion, as the script is, via the ACPI event, calling itself, which explains the flood of events, and probably also the other input-related problems.

As for the power button, I have no idea what's causing the reboot. You should be able to configure that via Gnome (or your DE), by default this should suspend the system. Does this also happen when you shut down without using the button? If so you could maybe try booting into Windows and see what happens when you shut down from there.

**Comment by deb75** [comment]
Thanks for your help,

I begin to understand that I have little understanding of this issue.

Shutting down without the button always worked. In windows, I think the button is configured to put the SB2 in suspend (connected standby) mode, that is why I never shutdown with the button in windows, in order to make sure the laptop is really off.

I do not catch where I can configure the power button action, I originally thought it was in the `/etc/acpi` directory. My DE is awesome, it is configured to call `/sbin/shutdown -h now` with a certain key press, but that is all.

**Comment by qzed** [comment]
So it properly shuts down when you don't use the button? For me this seems a bit like a configuration problem. I've just had a look at what could get notified by a power button press. Apparently that is systemd (`/etc/systemd/logind.conf`, `HandlePowerKey`) and the ACPI event files. Also I noticed that for me there was already an event file which is being triggered by the power button. Maybe one of them is configured to restart, or they do multiple conflicting actions?

Also I've just had an idea how we could test if this is a configuration problem. What happens if you run the following?
```shell
. /usr/share/acpi-support/key-constants
acpi_fakekey $KEY_POWER
```
This should act like you pressed the power button, so it's behaving the same way it should be a configuration issue.

**Comment by deb75** [comment]
Hi,

I tried `sudo acpi_fakekey 116` (`KEY_POWER=116`). It properly shuts down and remains off until I press the power button. Only odd thing is that I have to issue twice the command, once nothing happens, and the second time it begins the shutting down process.

The button power issue persists, when I press it, it shuts down and restart a few seconds after.

**Comment by deb75** [comment]
Got it.

The acpi scripts are bypassed by `systemd`. Once I set `HandlePowerKey` to `ignore`, the power button has no more effect. If I enter on the command line `systemctl poweroff`, I get the same issue : reboot after a few seconds, whereas `shutdow -h now` works fine.

So I created another acpi event for the power button with a script only containing `shutdown -h now` and it works well.

Thanks again for your help.

Regards

**Comment by qzed** [comment]
Awesome, meanwhile I managed to reproduce the issue by booting into the TTY instead of a window manager. Your fix seems to work.

I tried setting `HandlePowerKey` to `poweroff` which does not seem to have any effect for me, but running `systemctl poweroff` after pressing the button results in a reboot. This issue is really odd, I have absolutely no idea what's going on there. It's also weird that this does not happen for me with Gnome, I always thought Gnome would be using systemd under the hood. If anyone knows what's happening there, I'd appreciate your insight.

It's nice that you found a way to make this work! I suspect a few others will also run into this when running some of the more lighter DMs/WMs.

**Comment by qzed** [comment]
Yeah that's really a bit weird. I still suspect that somehow different handlers are getting called. I really don't see how the button press is different from a manually created power-button signal (everything the driver is doing is just mapping the GPIO interrupt to such a signal, by use of an already existing module).

Sadly I don't have much time at the moment, but I'll have a look at this once I do, probably about in a month or so.

**Comment by qzed** [comment]
I've had a look at this again, and I'm completely out of ideas now. First off, I removed all the event handlers in `/etc/acpi/events`, just to make sure that nothing there is interfering.
 
For _some_ reason, using the button (having `HandlePowerKey` set to `poweroff`) is behaving differently than powering off the device from the command line using `systemctl poweroff` _without ever touching the button_. With `HandlePowerKey` set to `ignore`, pressing the button does nothing, as it should, but shutting down the system after this using `systemctl poweroff` also causes a reboot.

I experience the same behavior using Gnome to shut down the device, and even `/sbin/shutdown -h now` (however the last one seems to work _sometimes_).

So, TLDR: Pressing the power-button has some side-effect that turns a shutdown into a reboot.

Maybe someone else has any idea what could be going on? I couldn't find any relevant information in the logs that would indicate that something is off. Perhaps ACPI could be doing some odd stuff?

Anyway, I'll keep on trying to figure this thing out, might just take longer than expected.

**Comment by qzed** [comment]
Ok, I think I'm getting a tiny bit closer to the root of this problem: When I unload the button driver (`modprobe -r surfacebook2_button`, confirmed with volume-buttons), then press the power-button and power off the system via `systemctl poweroff`, the system also reboots.

Can anyone confirm this?

Based on this I'd say there is something going on in the ACPI.

**Comment by qzed** [comment]
Next interesting thing: This does not happen on the mainline kernel.

**Comment by qzed** [comment]
The problem seems to occur first on the surface-kernel 4.17.1 release.

@jakeday Any idea?

**Comment by jakeday** [comment]
@qzed Does this still occur on 4.18.7-2?

**Comment by qzed** [comment]
No, seems to be working properly now, thanks!

**Comment by jakeday** [comment]
Cool! 😀
